mantidproject/mantid
# Running external local reconstruction

Description of work.

- Local reconstructions can now be launched from Mantid Imaging interface
  - On Linux only an actual reconstruction can be launched
  - On Windows we can still check if the external process runs, but most likely we'll get an import error
- The system settings weren't updated when the External Interpreter path and the Scripts location was being changed in the `System` tab -> `Execution of reconstruction jobs` -> `Local runs`
- Currently only a **single** reconstruction is possible, as running a second one _while_ the first is running will prompt the user and if accepted will terminate the first process
- Updated the reconstruction scripts to run on both Python 2 and 3 
  - However the tools themselves (TomoPy, Astra, Savu, etc) don't run on Python 3

Prerequisites:
- Python with TomoPy, I used [Anaconda2](https://www.continuum.io/downloads) for both Windows and Linux
- TomoPy v0.1.x is supported
  - This was tested locally with TomoPy v0.1.11 (installed with Anaconda, 2.7)
  - A temporary ["patch"](https://github.com/mantidproject/mantid/pull/18095/files#diff-85d4ffec21b1b51573f92591870ed3ecR619) has been added that allows TomoPy >1.x.x to run as well, however the results are not guaranteed to be accurate
- Test data `RB000888_test_stack_larmor_summed_201510` is available on `babylon5/Public/DimitarTasev` (~76.5 MB)


To test (both Windows and Linux):
-----------
- Build and open interface
- On the `Settings` tab check the external interpreter path, this will probably have just `python`, so make sure it points to the python that has `tomopy` installed
  - The external `python` process initialises it's path from the `PYTHONPATH` variable and it seems to ignore the `PATH` variable, so make sure the `tomopy` package is available in the `PYTHONPATH` variable
- Load a single image (or a stack, it doesn't matter as even a single image will set the paths in the `Setup` tab)
- Check that the paths in the `Setup` tab are correct
  - _Optional_: Go to the `ROI etc` tab and put some settings on (the settings don't matter for this PR)
- On the `Run` tab change resource to local and select `TomoPy` and click reconstruct (setup is optional, there are default values loaded with each tool)
  - This has only been tested with the `gridrec` algorithms (also the default one)
- Check that _any_ output is shown in the `Results Log`, as this should contain any errors as well as standard output, as long as the process started
  - Currently there is no error shown if the process isn't started, this is because Qt doesn't use an exception to signify that it failed to start
    - However there will be nothing added to the `Reconstruction jobs` in the `Run` tab, because no `started()` signal will be emitted from the external process and thus no job will be added 

Fixes #17940.



#### Release Notes:
[`.../release/imaging.rst`](https://github.com/mantidproject/mantid/pull/18095/files#diff-e678fb82fbdc92d52b938c69b0858a2e)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
